packages/@lwc/ssr-runtime/src/render.ts

```diff
@@ -39,22 +58,80 @@ export function* fallbackTmpl(
     }
 }
 
+export function fallbackTmplNoYield(
+    emit: (segment: string) => void,
+    _props: unknown,
+    _attrs: unknown,
+    _slotted: unknown,
+    Cmp: LightningElementConstructor,
+    _instance: unknown
+) {
+    if (Cmp.renderMode !== 'light') {
+        emit('<template shadowrootmode="open"></template>');
+    }
+}
+
 export type GenerateMarkupFn = (
     tagName: string,
     props: Record<string, any> | null,
     attrs: Attributes | null,
     slotted: Record<number | string, AsyncGenerator<string>> | null
 ) => AsyncGenerator<string>;
 
+export type GenerateMarkupFnAsyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => Promise<void>;
+
+export type GenerateMarkupFnSyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => void;
+
+type GenerateMarkupFnVariants =
+    | GenerateMarkupFn
+    | GenerateMarkupFnAsyncNoGen
+    | GenerateMarkupFnSyncNoGen;
+
 export async function serverSideRenderComponent(
     tagName: string,
-    compiledGenerateMarkup: GenerateMarkupFn,
-    props: Record<string, any>
+    compiledGenerateMarkup: GenerateMarkupFnVariants,
+    props: Record<string, any>,
+    mode: 'asyncYield' | 'async' | 'sync' = 'asyncYield'
 ): Promise<string> {
     let markup = '';
 
-    for await (const segment of compiledGenerateMarkup(tagName, props, null, null)) {
-        markup += segment;
+    if (mode === 'asyncYield') {
```

Comment:
We can sniff this, rather than needing a parameter. Dunno whether we should.

```js
// nb: would need to add a no-op for `emit` for async yield
const result = compiledGenerateMarkup(...);
if (typeof result?.next === 'function') {
  // async yield
} else if (result instanceof Promise) {
  // async
} else {
  // sync
}
```

Reply:
I like your other idea of baking the compilation mode into the generated code better. This would work but feels like an unnecessary hack when we can just be explicit.